web/yo/app/scripts/services/mainutils.js

```diff
@@ -664,6 +664,125 @@ angular.module('oncokbApp')
                 return item.indexOf("$") !== 0;
             });
         }
+        function checkDifferenceBetweenTherapies(oldContent, newContent){
+            var oldArray = _.flatten(therapyStrToArr(oldContent));
+            var newArray = _.flatten(therapyStrToArr(newContent));
+            var difference = {
+                'sameDrugs': _.intersection(oldArray, newArray),
+                'extraDrugsInOld': _.difference(oldArray, newArray),
+                'extraDrugsInNew': _.difference(newArray, oldArray)
+            };
+            return difference;
+        }
+
+        function changeMapByCurator(actionType, dataType, geneName, mutationUuid, mutationName, cancerTypeUuid, therapyUuid, content, oldContent){
+            switch (actionType) {
+                case 'add':
+                    changeMapWhenAdd(dataType, geneName, mutationUuid, mutationName, cancerTypeUuid, therapyUuid, null, content);
+                    break;
+                case 'name':
+                    var difference = checkDifferenceBetweenTherapies(oldContent, content);
+                    changeMapWhenRemove(dataType, geneName, mutationUuid, mutationName, cancerTypeUuid, therapyUuid, difference.extraDrugsInOld);
+                    changeTherapyNameInMapWithValidatingStatus(dataType, geneName, mutationUuid, mutationName, cancerTypeUuid, therapyUuid, difference.sameDrugs, content);
+                    changeMapWhenAdd(dataType, geneName, mutationUuid, mutationName, cancerTypeUuid, therapyUuid, difference.extraDrugsInNew, content);
+                    break;
+                case 'remove':
+                    var drugArray = _.flatten(therapyStrToArr(content));
+                    changeMapWhenRemove(dataType, geneName, mutationUuid, mutationName, cancerTypeUuid, therapyUuid, drugArray);
+                    break;
+            }
+        }
+
+        function changeMapWhenAdd(dataType, geneName, mutationUuid, mutationName, cancerTypeUuid, therapyUuid, drugs, content){
+            switch (dataType){
+                case 'mutation':
+                    break;
+                case 'tumor':
+                    break;
+                case 'treatment':
+                    var drugArray = [];
+                    if (!drugs){
+                        drugArray = _.flatten(therapyStrToArr(content));
+                    }
+                    else {
+                        drugArray = drugs;
+                    }
+                    var therapyObject = {
+                        'name': content,
+                        'status': 'latest'
+                    };
+                    var mapPath = '';
+                    var chainDrug = $q.when();
```

Comment:
If we enable typescript, the way you manipulating chainDrug is not acceptable. Rather than keep attaching to the chainDrug, why not just create an array and keep executing until the array is empty?